### DISCUSSION TITLE
Upcoming change: API Gateway to access the Validator

### DISCUSSION THREAD

**Discussion by MarcoMinghini** [INSPIRE-MIF/helpdesk-validator#769] [Announcements]
Dear all,
this message is to inform you about a change we will introduce in the INSPIRE Reference Validator in the next release (v.2021.3) planned for mid-September (see the [release plan](https://github.com/INSPIRE-MIF/helpdesk-validator/tree/master/release%20strategy)).

Starting with this release, access to the Validator will be managed and controlled by an API Gateway. While this will have no impact on users using the Validator from the INSPIRE central User Interface, users using the Validator API from external clients will need a token to access the API. 

More information on the access conditions as well as how organisations can request a token will follow here.

**Comment by MarcoMinghini** [comment]
Dear INSPIRE validation community,
As previously announced, the new release (v2021.3) of the INSPIRE Reference Validator will be accompanied by an additional security measure in the form of an API Gateway.

To give the community the opportunity to test their own workflows against the new API Gateway while also ensuring business continuity, after the release the current endpoint of the Validator will remain available for a **transition period of 3 weeks** (i.e. until October 6, 2021). After this period, the INSPIRE Reference Validator will be only accessible through the public endpoint of the API Gateway.

The API Gateway will impose a restriction on the rate and quota of the requests that any given user can send to the Validator. This limitations will be handled by the usage of API keys, which will be provided by the Validator team upon request from the organisations. The API key would need to be added as a header named `X-API-key` on the requests to the `/validator/v2/TestRuns` path of the API.

The system is already available at https://y9tuozgmmc.execute-api.eu-west-1.amazonaws.com/validator/v2/ for anyone who wants to perform tests and adapt their applications to the upcoming test.

Users and organisations interested in receiving an API key will need to contact the Validator team by sending an e-mail to `JRC-INSPIRE-SUPPORT AT ec.europa.eu` and provide the following information:
* name of the organisation
* responsible contact point and e-mail address (a personal e-mail address, in case the e-mail is sent from an organisation e-mail address)
* intended use of the API key

**Comment by ghost** [comment]
Hi @MarcoMinghini ,

thanks a lot for the information. 

Does the new API (v2021.3 - new UI) beside mentioned security improvements has some additional improvements/changes which affect clients (for example monitoring-bulk-validation-tool) which use API (v2021.1 - old UI)?  

For example: 
- performance improvement (validation is faster because API design has been changed - API caching)
- functional improvements (existing operation has been changed) 
- functional changes (new operation has been added)

**Comment by MarcoMinghini** [comment]
Dear all,
this message is to inform you that the transition period has been extended for an additional month to allow as many users as possible to request an API key and test it, while still being able to access the Validator without.

The public access to the Validator will be dismissed in the next days, which means that, in order to access the Validator from outside the public UI, you will need an API key. If you do not yet have one, you need to contact the Validator team by sending an e-mail to `JRC-INSPIRE-SUPPORT AT ec.europa.eu` and providing the following information:
* name of the organisation
* responsible contact point and e-mail address (a personal e-mail address, in case the e-mail is sent from an organisation e-mail address)
* intended use of the API key

Finally, I would like to inform the users of API keys that, after receiving some requests and in order to allow for a better preparation of the upcoming INSPIRE Monitoring and Reporting, the limitations for the API key usage have been relaxed. API key users will receive an update via e-mail.

**Comment by MarcoMinghini** [comment]
We have completed the migration to the API Gateway in the production instance. From now on, validations performed using the usual endpoint https://inspire.ec.europa.eu/validator/v2/ will get a 404 Not Found error. 

Once again, to use the Validator through the API (i.e. not from the UI) you will need to ask for an API key as described above.

**Comment by carlospzurita** [comment]
As an addendum to this announcement, we recommend users to clear any cached files on their browser before using the validator (see #667 ), as the user interface now should point to a new endpoint for the API Gateway.

**Comment by josegar74** [comment]
Hi

Is there any API service to retrieve the quotas available?

**Comment by jenriquesoriano** [comment]
Dear @josegar74  , 

we're checking the API docs in case there is a public endpoint ot retrieve the quotas.
We will keep you posted.

Best

**Comment by jenriquesoriano** [comment]
Dear @josegar74  , 

just an update.
We are analyzing how to make possible to retrieve the quota consistently using the API-Key via the usage method from the AWS rest services for the API gateway.
We will keep you posted.

Best,

**Comment by PaoloDiviacco** [comment]
Hi there.

I sent an email to get the Api-Key but received nothing ...what should I do?

Thank you in advance
P

**Comment by fabiovinci** [comment]
Dear @PaoloDiviacco,

we are replying to your email.

**Comment by PaoloDiviacco** [comment]
Thank you very much/Grazie mille

Paolo Diviacco
______________________________________________________

Dott. Paolo Diviacco, PhD
Sezione di Geofisica
Istituto Nazionale di Oceanografia e di Geofisica Sperimentale - OGS
Borgo Grotta Gigante n. 42/c
34010 Sgonico (Trieste) - Italia
Tel. +39 040 2140380
www.inogs.it

**Comment by PaoloDiviacco** [comment]
Hi there

Thank you for sending the key...seemingly it works, but unfortunately after a couple of submissions it replies that we have exceeded the monthly quota ...

Any suggestion?

Thank you 
Paolo Diviacco

![errorgeonetwork](https://user-images.githubusercontent.com/103563045/165253497-d3d0b10c-0faa-4497-ab65-e84a741cf440.png)

**Comment by fabiovinci** [comment]
Dear @PaoloDiviacco,

we checked the use of your API key and we detected that on April 13th you reached your weekly quota (200 requests).
This limit resets every Sunday at 00:00:00, so you should now be able to send new requests.

Could you try again?

**Comment by PaoloDiviacco** [comment]
Dear Fabio.

Thank you for your mail

This sounds very strange, we actually submitted only three records before being stopped.

Will check once more.

Anyway thank you for your reply.

Kind regards

Paolo
______________________________________________________

Dott. Paolo Diviacco, PhD
Sezione di Geofisica
Istituto Nazionale di Oceanografia e di Geofisica Sperimentale - OGS
Borgo Grotta Gigante n. 42/c
34010 Sgonico (Trieste) - Italia
Tel. +39 040 2140380
www.inogs.it

**Comment by PaoloDiviacco** [comment]
> Dear @PaoloDiviacco,
> 
> we checked the use of your API key and we detected that on April 13th you reached your weekly quota (200 requests). This limit resets every Sunday at 00:00:00, so you should now be able to send new requests.
> 
> Could you try again?

Dear Fabio... strange behaviour continues. We sent only few records, seemingly there is a loop of validation attempts...maybe these increase the counter and then eventually we are cut off after the limit you mention...but the actual submissions are very few. What can we do?

**Comment by fabiovinci** [comment]
Dear @PaoloDiviacco,

indeed it's very strange, if there is a loop of validation attempts it could be a problem related to Geonetwork, have you tried to open an issue in the related repository?

**Reply by fabiovinci** [reply]
An issue related to this error was opened in the GeoNetwork GitHub repository (https://github.com/geonetwork/core-geonetwork/issues/6463).

**Comment by dportoles** [comment]
I think Swagger docs https://inspire.ec.europa.eu/validator/swagger-ui.html (and probably other websites) are outdated because it doesn't explain details about how to get and use the new api key required.
Apart from that, e.g. 403 is not a documented return codes in expected response messages... and so on.
Could anybody update the docs? (I can open an issue if you want...)

**Comment by fabiovinci** [comment]
As announced [here](https://github.com/INSPIRE-MIF/helpdesk-validator/issues/1190), the API Gateway has been discontinued, so this discussion is closed as it is no longer relevant.
